### ISSUE TITLE
Difference in behavior between DAPR on local docker, vs DAPR inside .devcontainer (does not load components)

### ISSUE THREAD

**Issue by mattmazzola** [dapr/dapr#5133]
## In what area(s)?

/area runtime

## Ask your question here

I have created my own test application based on the `hello-world-compose` of Dapr samples. 
When I run `docker compose up` locally DAPR loads the statestore component for Redis correctly and all using the application is successful.

I then learned about docker .devcontainers and tried to add that to my sample.
The dev container seems to work, but when I run the same `docker compose up` inside the devcontainer it seems the statestore component is NOT loaded or discovered by the DAPR sidecar applications.

This means when I attempt the same usage of the application I receive
```
{\"errorCode\":\"ERR_STATE_STORE_NOT_CONFIGURED\",\"message\":\"state store is not configured\"}
```

I have created a video to demonstrate here:
https://drive.google.com/file/d/1fQWHjoUARUCTejcYyEiB1lgrUEvpSuKB/view?usp=sharing

I also have uploaded the sample used here:
https://github.com/mattmazzola/compose-network-test

To look at the devcontainer, docker-compose file, and the volume mounts

**Comment by mattmazzola** [comment]
Related to this issue, I would want to know, **what the best way to debug this type of issue?**

I have turned on Debug logging on both of the DAPR sidecars, but I didn't see any more useful information into the logs.
My next attempt to try to open an interactive shell inside the dapr container, but didn't seem to find any shell like executables.

Something like:
```
docker exec -it workspace-db-dapr-1 bash
```
Fails because `bash` or `sh` do not exist inside the sidecars.


```

**Comment by berndverst** [comment]
Can you just use our official dev container here? https://github.com/dapr/dapr/blob/master/.devcontainer/devcontainer.json

That should docker in docker -- which is probably what isn't working for you right now.

Maybe @ItalyPaleAle has more advice.

**Comment by ItalyPaleAle** [comment]
Hi @mattmazzola , I haven’t tested this personally, but at a quick look I can’t see you running `dapr init` In the dev container. Besides creating the redis container (managed by Dapr in this case), this also creates the config files and anything else that is needed.

Also, please don’t use `redis:alpine` because we don’t support version 8 of Redis yet.

As for using the dev containers from this repo, you may find that interesting but they are optimized for developing Dapr itself and not necessarily for developing Dapr-ized apps.

Lastly, Dapr production containers do not contain a shell by design, for security reasons.

**Comment by mattmazzola** [comment]
> Can you just use our official dev container here? https://github.com/dapr/dapr/blob/master/.devcontainer/devcontainer.json
> That should docker in docker -- which is probably what isn't working for you right now.

I agree with ItalyPaleAle that I shouldn't be using that image in this case

> As for using the dev containers from this repo, you may find that interesting but they are optimized for developing Dapr itself and not necessarily for developing Dapr-ized apps.

In other words, the devcontainer.json you linked seems intended for development of DAPR, but I don't intend to develop DAPR runtime. I want to develop an application which _uses_ DAPR runtime. The devcontainer.json was created by DAPR VScode extension so it should be "official".

> I can’t see you running dapr init In the dev container. Besides creating the redis container (managed by Dapr in this case), this also creates the config files and anything else that is needed.

I don't think `dapr init` is needed or would have effect. The devcontainer has instructions to run `init` so it should be ready for use:

see: https://github.com/microsoft/vscode-dev-containers/blob/b5dd6f89c03d57615b5e3fadd3c9ea572ca782c3/containers/dapr-javascript-node/.devcontainer/devcontainer.json#L30
```
	// Ensure Dapr is running on opening the container
	"postCreateCommand": "dapr uninstall --all && dapr init",
```

>  creating the redis container (managed by Dapr in this case), this also creates the config files and anything else that is needed.

Aside from init already being run. My understanding is Dapr init sets up default config to point to a redis instance so devs can easily get started using the dapr cli and dependencies already exist. However, this approach is not indented for more sophisticated development scenarios.

I don't want to rely on existing resources or DAPR defaults. I am running my own version of Redis (which is why it's on a different port) and using my own configuration .yml files, but those configuration files aren't loaded in the container hence the issue of not being able to use Redix

> we don’t support version 8 of Redis yet.

Ok, thanks. I didn't know that. I will adjust the Redis version and try again. 

> Dapr production containers do not contain a shell by design, for security reasons.

Ah, this makes sense and answers why "attempt to try to open an interactive shell inside the dapr container" did not work.

Although I would still like to know answer to the higher level question:

What the best way to debug when you confirm the services and sidecar containers running but the logs indicate configuration NOT being found?

These are things I think I should verify
1. How do I verify the components files are copied into the dapr sidecars?
1. How to compare file paths of where the DARP sidecar is attempting to load components vs where the components files are located? (I should be the `--components-path` location)

**Comment by ItalyPaleAle** [comment]
Take a look at using Dapr with Docker-compose: https://docs.dapr.io/operations/hosting/self-hosted/self-hosted-with-docker/#run-using-docker-compose

**Comment by mattmazzola** [comment]
I wasn't sure which part you wanted to look at. I think your link on shows a snippet or single file of the "hello-world-compose" sample. In my first post I said I based my sample off of [this sample](https://github.com/dapr/samples/tree/master/hello-docker-compose)

I would bet the issue still remains. I think we could test that the `hello-world-compose` sample works locally but does not work on the devcontainer.

Also, from these [docs](https://docs.dapr.io/concepts/dapr-services/placement/#:~:text=The%20Dapr%20Placement%20service%20is%20used%20to%20calculate,actors%20running%20in%20self-hosted%20mode%20or%20on%20Kubernetes.): 

> The Dapr Placement service is used to calculate and distribute distributed hash tables for the location of [Dapr actors](https://docs.dapr.io/developing-applications/building-blocks/actors/)

and because the sample isn't using actors, we would not need the placement service or parameters to sidecars. 
I thought we could keep this on issue of loading components

**Comment by mattmazzola** [comment]
Is there any more suggestions on how to debug this issue?
Does anyone have answers to the questions above:
1. How do I verify the components files are copied into the dapr sidecars?
1. How to compare file paths of where the DARP sidecar is attempting to load components vs where the components files are located? (I should be the --components-path location)

Since I have linked the repo I think it should be fairly easy for someone to clone and attempt to reproduce the behavior.
If someone else observes the same error, then it is likely issue with the .devcontainer

Or likewise if someone has a DAPR repo setup with .devcontainer using docker compose that I could try. If I could load components using .devcontainer in another repo, then I would know the problem lies with my repo.

**Comment by berndverst** [comment]
@ItalyPaleAle can you add the "pinned" label so this doesn't get closed? (And maybe you know what to do for this issue)

**Comment by eyenx** [comment]
> we don’t support version 8 of Redis yet.

@ItalyPaleAle any roadmap on this?
